File: tornado_bench/bench.py

import sys
import time
import tornado.ioloop
import tornado.gen
import time
from nats.io.utils import new_inbox
from nats.io.client import Client as NATS

subject = "0123456789012345"
payload = "0123456789012345012345678901234501234567890123450123456789012345"


class Client(object):
    
    def __init__(self, nc):
        self.nc = nc
        self.total = 0
        
    def connect(self, **kwargs):
        return self.nc.connect(**kwargs)
        
    def on_receive(self, msg=None):
        self.total += 1
        
    def subscribe(self, subject):
        return self.nc.subscribe(subject, '', self.on_receive)
    
    def publish(self, subject, payload):
        self.total += 1
        return self.nc.publish(subject, payload)
        

@tornado.gen.coroutine
def main():
    if len(sys.argv) != 3:
        print 'Usage: python bench.py nats://HOST:PORT message_count'
        sys.exit(1)
    url = sys.argv[1]
    msg_count = int(sys.argv[2])
    
    pub = Client(NATS())
    yield pub.connect(verbose=False, servers=[url])
    
    sub = Client(NATS())
    yield sub.connect(verbose=True, servers=[url])
    yield sub.subscribe(subject)
    
    tic = time.time()
    for i in xrange(msg_count):
        yield pub.publish(subject, payload)
    tac = time.time()
    
    yield tornado.gen.sleep(1)
    
    secs = tac - tic
    usec = secs * 1000000
    ops = msg_count / secs
    print msg_count, int(usec), int(round(ops))

if __name__ == '__main__':
    tornado.ioloop.IOLoop.instance().run_sync(main)

File: elixir_bench/bench.exs
defmodule Bench do    
    @subject "0123456789012345"
    @payload "0123456789012345012345678901234501234567890123450123456789012345"
    
    def go(nats_conf, msg_count) do
        subject = @subject
        payload = @payload
        start_subscribe(self, nats_conf, subject, msg_count)
        :timer.sleep(1000)  # Ensure we are subscribed
        f = fn ->
            publish_and_receive(nats_conf, subject, payload, msg_count)
        end
        {usec, :ok} = :timer.tc(f)
        ops = ops(usec, msg_count)
        {msg_count, usec, ops}
    end
    
    defp publish_and_receive(nats_conf, subject, payload, msg_count) do
        start_publish(nats_conf, subject, payload, msg_count)
        receive do
            :done -> :ok
        end
    end
    
    defp ops(usec, msg_count) do
        round(msg_count / (usec / 1_000_000))
    end
    
    def subscribe(parent, sub, subject, msg_count) do
        Nats.Client.sub(sub, self, subject)
        loop(sub, msg_count)
        send parent, :done
    end

    defp loop(_, 0) do
        :ok
    end
    
    defp loop(conn, msg_count) do
        receive do
            {:msg, _, _, _, _} ->
                loop(conn, msg_count - 1)
        end
    end
    
    def publish(_, _, _, 0) do
        :ok
    end

    def publish(conn, subject, payload, left) do
        Nats.Client.pub(conn, subject, payload)
        publish(conn, subject, payload, left - 1)
    end
    
    def start_subscribe(parent, nats_conf, subject, msg_count) do
        {:ok, sub} = Nats.Client.start_link(nats_conf)
        spawn(fn -> subscribe(parent, sub, subject, msg_count) end)
    end
    
    def start_publish(nats_conf, subject, payload, msg_count) do
        {:ok, pub} = Nats.Client.start_link(nats_conf)
        spawn(fn -> publish(pub, subject, payload, msg_count) end)
    end
end

defmodule ArgvParser do
    def nats_conf do
        <<"nats://", host_port :: binary>> = :lists.nth(1, System.argv)
        [host, bin_port] = :binary.split(host_port, ":")
        port = :erlang.binary_to_integer(bin_port)
        %{host: host, port: port}
    end
    
    def message_count do
        :erlang.binary_to_integer(:lists.nth(2, System.argv))
    end

end

if length(System.argv) == 2 do
    {msg_count, usec, ops} = Bench.go(ArgvParser.nats_conf,
                                      ArgvParser.message_count)
    IO.puts "#{msg_count} #{usec} #{ops}"
else
    IO.puts "Usage: iex bench.exs nats://HOST:PORT message_count"
end

